Sources/AsyncHTTPClient/Utils.swift

```diff
@@ -77,149 +78,207 @@ extension NIOClientTCPBootstrap {
-            channel = bootstrap.connect(host: address.host, port: address.port)
-        case .unix, .http_unix, .https_unix:
-            channel = bootstrap.connect(unixDomainSocketPath: key.unixPath)
-        }
+        let bootstrap = NIOClientTCPBootstrap.makeHTTPClientBootstrapBase(on: channelEventLoop,
+                                                                          host: key.host,
+                                                                          port: key.port,
+                                                                          requiresTLS: requiresTLS,
+                                                                          configuration: configuration,
+                                                                          sslContextCache: sslContextCache,
+                                                                          logger: logger)
+        return bootstrap.flatMap { bootstrap -> EventLoopFuture<Channel> in
+            let channel: EventLoopFuture<Channel>
+            switch key.scheme {
+            case .http, .https:
+                let address = HTTPClient.resolveAddress(host: key.host, port: key.port, proxy: configuration.proxy)
+                channel = bootstrap.connect(host: address.host, port: address.port)
+            case .unix, .http_unix, .https_unix:
+                channel = bootstrap.connect(unixDomainSocketPath: key.unixPath)
+            }
 
-        return channel.flatMap { channel in
-            let requiresTLS = key.scheme.requiresTLS
-            let requiresLateSSLHandler = configuration.proxy != nil && requiresTLS
-            let handshakeFuture: EventLoopFuture<Void>
-
-            if requiresLateSSLHandler {
-                let handshakePromise = channel.eventLoop.makePromise(of: Void.self)
-                channel.pipeline.syncAddLateSSLHandlerIfNeeded(for: key, tlsConfiguration: configuration.tlsConfiguration, handshakePromise: handshakePromise)
-                handshakeFuture = handshakePromise.futureResult
-            } else if requiresTLS {
-                do {
-                    handshakeFuture = try channel.pipeline.syncOperations.handler(type: TLSEventsHandler.self).completionPromise.futureResult
-                } catch {
-                    return channel.eventLoop.makeFailedFuture(error)
+            return channel.flatMap { channel -> EventLoopFuture<(Channel, NIOSSLContext?)> in
+                sslContext.map { sslContext -> (Channel, NIOSSLContext?) in
+                    (channel, sslContext)
+                }
+            }.flatMap { channel, sslContext in
+                configureChannelPipeline(channel,
+                                         isNIOTS: bootstrap.isNIOTS,
+                                         sslContext: sslContext,
+                                         configuration: configuration,
+                                         key: key)
```

Comment:
NIT: Do we maybe collapse this into one closure?

Reply:
which closures do you want me to collapse?

Reply:
Nevermind. I thought we can collapse the `sslContext.map` into the following `flatMap`, but that doesn't work, since the `sslContext` is a `EventLoopFuture<NIOSSLContext?>` first and actual `NIOSSLContext?` later.